storage-proofs/src/rational_post.rs

```diff
@@ -0,0 +1,473 @@
+    }
+
+    #[test]
+    fn rational_post_blake2s() {
+        test_rational_post::<Blake2sHasher>();
+    }
+
+    // Construct a proof that satisfies a cursory validation:
+    // Data and proof are minimally consistent.
+    // Proof root matches that requested in public inputs.
+    // However, note that data has no relationship to anything,
+    // and proof path does not actually prove that data was in the tree corresponding to expected root.
+    fn make_bogus_proof<H: Hasher>(
+        pub_inputs: &PublicInputs<H::Domain>,
+        rng: &mut XorShiftRng,
+    ) -> MerkleProof<H> {
+        let bogus_leaf: H::Domain = rng.gen();
+        let hashed_leaf = H::Function::hash_leaf(&bogus_leaf);
+
+        make_proof_for_test(
+            pub_inputs.commitments[0],
+            hashed_leaf,
+            vec![(hashed_leaf, true)],
+        )
+    }
+
+    fn test_rational_post_validates<H: Hasher>() {
+        let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+
+        let leaves = 32;
+        let sector_size = leaves * 32;
+        let challenges_count = 2;
+        let pub_params = PublicParams {
+            sector_size,
+            challenges_count,
+        };
+
+        let data: Vec<u8> = (0..leaves)
+            .flat_map(|_| fr_into_bytes::<Bls12>(&rng.gen()))
+            .collect();
+
+        let graph = BucketGraph::<H>::new(32, 5, 0, new_seed());
+        let tree = graph.merkle_tree(data.as_slice()).unwrap();
+        let seed = (0..32).map(|_| rng.gen()).collect::<Vec<u8>>();
+        let faults = vec![];
+        let challenges = derive_challenges(challenges_count, sector_size, 2, &seed, &faults);
+        let commitments = challenges.iter().map(|_c| tree.root()).collect::<Vec<_>>();
+
+        let pub_inputs = PublicInputs::<H::Domain> {
+            challenges: &challenges,
+            faults: &faults,
+            commitments: &commitments,
+        };
+
+        let bad_proof = Proof {
+            inclusion_proofs: vec![
+                make_bogus_proof::<H>(&pub_inputs, rng),
+                make_bogus_proof::<H>(&pub_inputs, rng),
+            ],
+        };
+
+        let verified = RationalPoSt::verify(&pub_params, &pub_inputs, &bad_proof)
+            .expect("verification failed");
+
+        // A bad proof should not be verified!
+        assert!(!verified);
+    }
+
+    #[test]
+    fn rational_post_actually_validates_sha256() {
+        test_rational_post_validates::<Sha256Hasher>();
+    }
+
+    #[test]
+    fn rational_post_actually_validates_blake2s() {
+        test_rational_post_validates::<Blake2sHasher>();
+    }
+
+    #[test]
+    fn rational_post_actually_validates_pedersen() {
+        test_rational_post_validates::<PedersenHasher>();
+    }
+
+    fn test_rational_post_validates_challenge_identity<H: Hasher>() {
+        let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
```

Comment:
More magic values. Where do these things come from? What do they represent?

Reply:
see above

Reply:
See comment above. Requesting that you factor all of the `[0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]` into a constant.

Reply:
not in this pr, this would be a very independent change

Reply:
not in this pr, this would be a very independent change